src/vscode-dts/vscode.proposed.contribEditSessions.d.ts

```diff
@@ -3,4 +3,31 @@
  *  Licensed under the MIT License. See License.txt in the project root for license information.
  *--------------------------------------------------------------------------------------------*/
 
-// empty placeholder declaration for the `contribEditSessions`-contribution point
+declare module 'vscode' {
+
+	// https://github.com/microsoft/vscode/issues/157734
+
+	export namespace workspace {
```

Comment:
This introduces a concept of "canonical workspace identity" but the actual identity is expressed as `[key: string]: string | null` which confuses me at first. If this is about identity, I would have expected a UUID or `Uri` that I can use to compare with another workspace?

This looks to me more like a way to provide generic metadata for a workspace by an extension? Maybe it would help me if I were to understand how this metadata is used in our code, i.e. what drives this?

Reply:
The identity is associated with an edit session: https://github.com/microsoft/vscode/blob/ed4926377df73a74c45f52ae977223aa0e2041e5/src/vs/workbench/contrib/editSessions/browser/editSessions.contribution.ts#L446-L449 and used to determine whether a given edit session can be applied to the current workspace (currently just doing object property equality): https://github.com/microsoft/vscode/blob/ed4926377df73a74c45f52ae977223aa0e2041e5/src/vs/workbench/contrib/editSessions/browser/editSessions.contribution.ts#L341-L349

The URI is not quite sufficient to capture context about the current state, because for example if I have an edit session created for microsoft/vscode on main but a different SHA than I currently have open, the edit session is a partial match and we have reduced confidence that we can apply it without asking. Having the identity be expressed as `string` would also work, but providers would need to `JSON.stringify` the metadata before handing it out to VS Code (and know how to deserialize the metadata). In future I'd also like to introduce a provider method to express confidence about the match, e.g.

```ts

enum MatchType {
    Complete,
    Partial,
    None
}

export interface CanonicalWorkspaceIdentityProvider {
    /**
     *
     * @param workspaceFolder A workspace folder that this provider can provide a canonical workspace identity for.
     * @param canonicalIdentity A canonical identifier to match against the identity for the @param workspaceFolder.
     * @returns Whether the @param canonicalIdentity matches the @param workspaceFolder.
     */
    provideCanonicalWorkspaceIdentityMatchType(workspaceFolder: WorkspaceFolder, canonicalIdentity: string): MatchType

    /**
     *
     * @param workspaceFolder A workspace folder that this provider can provide a canonical workspace identity for.
     * @param token A cancellation token for the request.
     * @returns A string representing the canonical workspace identity for the requested workspace folder.
     */
    provideCanonicalWorkspaceIdentity(workspaceFolder: WorkspaceFolder, token: CancellationToken): ProviderResult<string>;
}
```

Reply:
I think my confusion originates from the fact that this talks about "canonical workspace identity" but I see this more as an identifier of a session where I worked in the workspace. Is it not possible that I may have been working on the same workspace in different machines in different edit sessions? If I am not connected to the internet, I may actually have accumulated more than one edit session on a workspace and thus we could have multiple canonical identities, where imho this terminology does not work anymore.

I wonder if we should change this API to talk about "edit sessions" then, where you could have more than one. If I have 2 machines that I worked on a workspace without internet, then turn on both internets and let them roam the "edit session", I will have 2 edit sessions. On my third machine, I want to continue one of the two sessions, so I will probably have to pick which one from which machine. Should this API then be an array of edit sessions and their identifiers?

Reply:
I just discovered that we already have a `CanonicalWorkspace` concept in internals here https://github.com/microsoft/vscode/blob/8e4ebf9f86c936b6ff28f85c6db169154fd6da17/src/vs/workbench/services/workspaces/common/workspaceTrust.ts#L34-L64 so to avoid confusion I will rename to `EditSessionIdentityProvider`.